Fix the bootloader

malec1984 opened the pull request:
Unfortunately the bootloader was broken after warnings removal patch.
Also took the opportunity to remove the unused files.

Please see the comments on the patch for the important changes, this in case you decide not to take this patch and fix with your commit.

malec1984 commented on software/front/US_BootLoader.X/PIC32MX564F128H.c:
```diff
@@ -141,17 +142,17 @@ UINT mcuWriteFlashRecord(void * RecordData)
-    UINT Result;
-    lFR.Address = ((UINT32 *)RecordData)[0];
-    lFR.RecDataLen = ((UINT32 *)RecordData)[1];
-    lFR.Data = (UINT8*)&((UINT32 *)RecordData)[3];
```
Here the important part is lFR.Data = (UINT8*)&((UINT32 *)RecordData)[2];
I've made it in the previous format as I find it easier to read.

sparkybg replied:
IMHO, indexing in square braces (addr[index]) is more readable than indexing with math (addr+index), despite that the effect is the same. It seems to me that square braces makes it more obvious that the indexing depends on data size than math.

The real problem here was that:
lFR.Data = (UINT8*)&((UINT32 *)RecordData)[3];
should have been:
lFR.Data = (UINT8*)&((UINT32 *)RecordData)**[2]**;

And I think this is the thing that broke the bootloader.

Is this editable? I really want to keep the square brace indexing instead of pointer math here.

malec1984 replied:
No, I'll have to upload a second patch for it.

malec1984 replied:
Done.

sparkybg replied:
Merged it. Look OK, but please verify that it works. I won't have time for this in the following day or two.

malec1984 commented on software/front/US_BootLoader.X/io.h:
```diff
@@ -49,17 +62,15 @@ typedef union {
-IOC_EXTERN volatile USBPacket RXP;
-IOC_EXTERN volatile USBPacket TXP;
+/*Buffers used for USB communication*/
+IOC_EXTERN USBPacket RXP;
+IOC_EXTERN USBPacket TXP;
```
The buffers used for USB communication should not be declared as volatile. The buffer is not modified in any way by the HW so it seems by using volatile is leading to strange things.

sparkybg replied:
From the XC32 manual:
..and any variables which may be modified by interrupt routines should use
this qualifier as well. For example:

USB routines can run in interrupt mode (although at the moment it does not). And I don't know if these buffers are modified by hardware (DMA or something else) or not.

The USB stack used is not only for this project - I may use it in the future (if not already, I don't remember) for another projects in interrupt mode.

malec1984 replied:
Well I have not yet found the reason why this is not working with volatile.
Your define was actually using the buffer USBTxBuffer 
#define TXP (*((USBPacket *)USBTxBuffer))

which is declared in usb_driver.c as unsigned char USBRxBuffer[64]

sparkybg replied:
Also:
/*memory move necessary because the command is on 1 byte, could be avoided by using 4 bytes*/
Yes, it can be avoided, but we will lose 3 bytes of data and the PC software must be edited to support this. The bootloader uses USB HID, and the pockets there are 64 bytes if I remember well.

I will think of a better way to do this somehow.

malec1984 replied:
Yes, I've looked on it too, that's why I only added the comment just to remember why that memmove is necessary.
The PC app I've seen and you have used just binary array to fill in the data, as it looks there are too many changes to make the command on 4 bytes

malec1984 reviewed (commented):
See the comments

malec1984 commented:
I rebased now and recompiled, flashed and updated firmware, works as expected :)
The only thing is that I do not commit the hex file, only the source code, maybe it would be better to put the hex only in releases, just an idea.
    On Sunday, January 17, 2021, 04:54:01 PM GMT+1, Spartak Timchev <notifications@github.com> wrote:  
 
 


@sparkybg commented on this pull request.

In software/front/US_BootLoader.X/PIC32MX564F128H.c:
> @@ -141,17 +142,17 @@ UINT mcuWriteFlashRecord(void * RecordData)
     void* ProgAddress;
     void* ProgAddressEnd;
     UINT32 WrData;
-    UINT Result;
-    lFR.Address = ((UINT32 *)RecordData)[0];
-    lFR.RecDataLen = ((UINT32 *)RecordData)[1];
-    lFR.Data = (UINT8*)&((UINT32 *)RecordData)[3];

Merged it. Look OK, but please verify that it works. I won't have time for this in the following day or two.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

sparkybg commented:
I will commit HEXes. There are several links to them already so some people will be confused if I stop publishing them. I will make a new release in a day or two. I think it will be better if we leave HEXes as they were unitil now.
